File: src/instance/lsfg.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/pelletier/go-toml/v2"
)

// LSFG Configuration Types

// LsfgGlobalConfig represents global LSFG-VK frame generation settings
type LsfgGlobalConfig struct {
	Version   int    `toml:"version"`
	AllowFP16 bool   `toml:"allow_fp16"`
	DLL       string `toml:"dll"`
}

// LsfgConfigProfile represents a per-game LSFG-VK frame generation profile
type LsfgConfigProfile struct {
	Name            string      `toml:"name"`
	ActiveIn        interface{} `toml:"active_in"` // Can be string or []string
	Multiplier      int         `toml:"multiplier"`
	PerformanceMode bool        `toml:"performance_mode"`
	GPU             string      `toml:"gpu"`
	FlowScale       float32     `toml:"flow_scale"`
	Pacing          string      `toml:"pacing"`
}

// LsfgConfigFile represents the complete LSFG-VK configuration file
type LsfgConfigFile struct {
	Version  int                 `toml:"version"`
	Global   LsfgGlobalConfig    `toml:"global"`
	Profiles []LsfgConfigProfile `toml:"profile"`
}

// LSFG Configuration Functions

// GetLsfgConfigPath returns the path to the lsfg-vk config file.
// By default lsfg-vk reads from ~/.config/lsfg-vk/conf.toml
func GetLsfgConfigPath() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(home, ".config", "lsfg-vk", "conf.toml"), nil
}

// FindLsfgProfileForGame finds the LSFG profile that applies to the given game path
func FindLsfgProfileForGame(gamePath string) (*LsfgConfigProfile, int, error) {
	configPath, err := GetLsfgConfigPath()
	if err != nil {
		return nil, -1, err
	}
	return findLsfgProfileForGameAtPath(gamePath, configPath)
}

// findLsfgProfileForGameAtPath finds the profile for a game at a specific config path
func findLsfgProfileForGameAtPath(gamePath, configPath string) (*LsfgConfigProfile, int, error) {
	data, err := os.ReadFile(configPath)
	if err != nil {
		return nil, -1, fmt.Errorf("failed to read LSFG config: %w", err)
	}

	var config LsfgConfigFile
	if err := toml.Unmarshal(data, &config); err != nil {
		return nil, -1, fmt.Errorf("failed to parse LSFG config: %w", err)
	}

	exeName := strings.ToLower(filepath.Base(gamePath))

	// Find the matching profile
	for i, profile := range config.Profiles {
		if matchesProfile(exeName, profile.ActiveIn) {
			return &profile, i, nil
		}
	}

	return nil, -1, fmt.Errorf("no matching LSFG profile found for %s", exeName)
}

// matchesProfile checks if the executable name matches a profile's active_in configuration
func matchesProfile(exeName string, activeIn interface{}) bool {
	exeName = strings.ToLower(exeName)

	switch v := activeIn.(type) {
	case string:
		return strings.EqualFold(v, exeName)
	case []interface{}:
		for _, item := range v {
			if s, ok := item.(string); ok {
				if strings.EqualFold(s, exeName) {
					return true
				}
			}
		}
	}
	return false
}

// EnsureLsfgProfileExists creates a profile for the game if one doesn't already exist
// Uses provided options for initial configuration, or defaults if not provided
func EnsureLsfgProfileExists(gamePath string, opts LaunchOptions) error {
	configPath, err := GetLsfgConfigPath()
	if err != nil {
		return fmt.Errorf("failed to get LSFG config path: %w", err)
	}

	// Try to find existing profile
	_, _, err = findLsfgProfileForGameAtPath(gamePath, configPath)
	if err == nil {
		// Profile already exists
		return nil
	}

	// Profile doesn't exist, create it
	DebugLog(fmt.Sprintf("Creating new LSFG profile for %s", gamePath))

	exeName := filepath.Base(gamePath)

	// Read existing config or create new one
	var config LsfgConfigFile
	data, err := os.ReadFile(configPath)
	if err != nil {
		// Config doesn't exist, create new
		config = LsfgConfigFile{
			Version:  2,
			Global:   LsfgGlobalConfig{Version: 2, AllowFP16: true, DLL: opts.LsfgDllPath},
			Profiles: []LsfgConfigProfile{},
		}
	} else {
		// Config exists, try to parse it
		if err := toml.Unmarshal(data, &config); err != nil {
			// Parsing failed, start fresh
			config = LsfgConfigFile{
				Version:  2,
				Global:   LsfgGlobalConfig{Version: 2, AllowFP16: true, DLL: opts.LsfgDllPath},
				Profiles: []LsfgConfigProfile{},
			}
		} else {
			// Successfully parsed - preserve existing global config, only update DLL if provided
			if opts.LsfgDllPath != "" {
				config.Global.DLL = opts.LsfgDllPath
			}
		}
	}

	// Parse multiplier from options
	multiplier := 2
	if opts.LsfgMultiplier != "" {
		if m, err := fmt.Sscanf(opts.LsfgMultiplier, "%d", &multiplier); m == 1 && err == nil {
			// Parsing succeeded
		}
	}

	// Create profile name from executable name (without extension)
	profileName := strings.TrimSuffix(exeName, filepath.Ext(exeName))

	// Parse flow scale from options
	flowScale := 1.0
	if opts.LsfgFlowScale != "" {
		if f, err := strconv.ParseFloat(opts.LsfgFlowScale, 32); err == nil {
			flowScale = f
		}
	}

	// Create new profile with values from options
	newProfile := LsfgConfigProfile{
		Name:            profileName,
		ActiveIn:        exeName,
		Multiplier:      multiplier,
		PerformanceMode: opts.LsfgPerfMode,
		GPU:             opts.LsfgGpu,
		FlowScale:       float32(flowScale),
		Pacing:          opts.LsfgPacing,
	}

	config.Profiles = append(config.Profiles, newProfile)

	// Ensure config directory exists
	if err := os.MkdirAll(filepath.Dir(configPath), 0755); err != nil {
		return fmt.Errorf("failed to create LSFG config directory: %w", err)
	}

	// Write config
	data, err = toml.Marshal(&config)
	if err != nil {
		return fmt.Errorf("failed to marshal LSFG config: %w", err)
	}

	if err := os.WriteFile(configPath, data, 0644); err != nil {
		return fmt.Errorf("failed to write LSFG config: %w", err)
	}

	DebugLog(fmt.Sprintf("Successfully created LSFG profile for %s", exeName))
	return nil
}

File: src/electron/launcher/lsfg.ts
import fs from "fs";
import path from "path";
import crypto from "crypto";
import os from "os";
import { getBaseDir } from "./utils.js";
import type { LsfgConfigFile, LsfgConfigProfile } from "./types.js";

export function getLsfgProfilePath(gamePath: string): string {
	const hash = crypto.createHash("sha1").update(gamePath).digest("hex").slice(0, 8);
	const exeName = path.basename(gamePath);
	const baseName = exeName.replace(/\.exe$/i, "");
	const filename = `${baseName}-${hash}.toml`;
	const configDir = path.join(getBaseDir(), "config", "lsfg");
	return path.join(configDir, filename);
}

export function getLsfgConfigPath(): string {
	return path.join(os.homedir(), ".config", "lsfg-vk", "conf.toml");
}

// Simple TOML Parser for the specific LSFG structure
function parseLsfgToml(data: string): LsfgConfigFile {
	const lines = data.split("\n");
	const config: any = { version: 2, global: {}, profile: [] };
	let currentSection = "";
	let currentProfile: any = null;

	for (let line of lines) {
		line = line.trim();
		if (!line || line.startsWith("#")) continue;

		if (line.startsWith("[") && line.endsWith("]")) {
			currentSection = line.slice(1, -1);
			if (currentSection === "profile") {
				if (currentProfile) config.profile.push(currentProfile);
				currentProfile = {};
			}
			continue;
		}

		const [key, ...valParts] = line.split("=");
		if (!key) continue;
		const val = valParts.join("=").trim();
		const k = key.trim();

		let parsedVal: any = val;
		if (val === "true") parsedVal = true;
		else if (val === "false") parsedVal = false;
		else if (val.startsWith('"') && val.endsWith('"')) parsedVal = val.slice(1, -1);
		else if (!isNaN(Number(val))) parsedVal = Number(val);
		else if (val.startsWith("[") && val.endsWith("]")) {
			// Simple array parser for active_in
			parsedVal = val
				.slice(1, -1)
				.split(",")
				.map((s) => s.trim().replace(/^"|"$/g, ""))
				.filter((s) => s);
		}

		if (currentSection === "global") {
			config.global[k] = parsedVal;
		} else if (currentSection === "profile") {
			currentProfile[k] = parsedVal;
		} else {
			config[k] = parsedVal;
		}
	}
	if (currentProfile) config.profile.push(currentProfile);

	return config as LsfgConfigFile;
}

function stringifyLsfgToml(config: LsfgConfigFile): string {
	let output = `version = ${config.version}\n\n`;
	output += `[global]\n`;
	output += `version = ${config.global.version}\n`;
	output += `allow_fp16 = ${config.global.allow_fp16}\n`;
	output += `dll = "${config.global.dll}"\n\n`;

	for (const p of config.profile) {
		output += `[[profile]]\n`;
		output += `name = "${p.name}"\n`;
		if (Array.isArray(p.active_in)) {
			output += `active_in = [${p.active_in.map((s) => `"${s}"`).join(", ")}]\n`;
		} else {
			output += `active_in = "${p.active_in}"\n`;
		}
		output += `multiplier = ${p.multiplier}\n`;
		output += `performance_mode = ${p.performance_mode}\n`;
		output += `gpu = "${p.gpu}"\n`;
		output += `flow_scale = ${p.flow_scale.toFixed(1)}\n`;
		output += `pacing = "${p.pacing}"\n\n`;
	}
	return output;
}

export function findLsfgProfileForGame(gamePath: string): { profile: LsfgConfigProfile | null; index: number } {
	const configPath = getLsfgConfigPath();
	if (!fs.existsSync(configPath)) return { profile: null, index: -1 };

	try {
		const data = fs.readFileSync(configPath, "utf8");
		const config = parseLsfgToml(data);
		const exeName = path.basename(gamePath).toLowerCase();

		for (let i = 0; i < config.profile.length; i++) {
			const p = config.profile[i];
			const activeIn = Array.isArray(p.active_in) ? p.active_in : [p.active_in];
			if (activeIn.some((s) => s.toLowerCase() === exeName)) {
				return { profile: p, index: i };
			}
		}
	} catch (e) {
		console.error("Failed to parse LSFG config:", e);
	}
	return { profile: null, index: -1 };
}

export function saveLsfgProfile(
	gamePath: string,
	opts: Partial<LsfgConfigProfile & { allowFp16: boolean; dllPath: string }>,
): void {
	const configPath = getLsfgConfigPath();
	const configDir = path.dirname(configPath);
	if (!fs.existsSync(configDir)) fs.mkdirSync(configDir, { recursive: true });

	let config: LsfgConfigFile = {
		version: 2,
		global: { version: 2, allow_fp16: opts.allowFp16 ?? true, dll: opts.dllPath ?? "" },
		profile: [],
	};

	if (fs.existsSync(configPath)) {
		try {
			config = parseLsfgToml(fs.readFileSync(configPath, "utf8"));
		} catch (e) {
			console.error("Failed to read existing LSFG config, creating new one");
		}
	}

	// Update global
	if (opts.allowFp16 !== undefined) config.global.allow_fp16 = opts.allowFp16;
	if (opts.dllPath !== undefined) config.global.dll = opts.dllPath;

	const exeName = path.basename(gamePath);
	const { index } = findLsfgProfileForGame(gamePath);

	const newProfile: LsfgConfigProfile = {
		name: exeName.replace(/\.[^/.]+$/, ""),
		active_in: exeName,
		multiplier: opts.multiplier ?? 2,
		performance_mode: opts.performance_mode ?? false,
		gpu: opts.gpu ?? "auto",
		flow_scale: opts.flow_scale ?? 1.0,
		pacing: "none",
	};

	if (index !== -1) {
		config.profile[index] = { ...config.profile[index], ...newProfile };
	} else {
		config.profile.push(newProfile);
	}

	fs.writeFileSync(configPath, stringifyLsfgToml(config));
}
